sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala

```diff
@@ -253,16 +253,94 @@ private[sql] object ArrowConverters extends Logging {
         val vectorLoader = new VectorLoader(root)
         vectorLoader.load(arrowRecordBatch)
         arrowRecordBatch.close()
+        vectorSchemaRootToIter(root)
+      }
+    }
+  }
+
+  /**
+   * Maps iterator from serialized ArrowRecordBatches to InternalRows. Different from
+   * [[fromBatchIterator]], each input arrow batch starts with the schema.
+   */
+  private[sql] def fromBatchWithSchemaIterator(
```

Comment:
Sorry for late reviews. Can we dedup the logic like `ArrowBatchWithSchemaIterator` is doing?